lib/install/management/configure.go

```diff
@@ -120,6 +126,8 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
 		}
 	}
 
+	err = d.update(conf, settings, isConfigureOp)
```

Comment:
I think the reason that this happens _before_ is that there's no way to rollback `opsuser.GrantOpsUserPerms` if this fails.

Reply:
To summarize this issue:

Today, configure (including upgrade) is organized in the following way:
1. A new configuration for the VCH is constructed based on the old configuration
   and requested changes.
2. The new configuration is validated.
3. The updated version of any ISO files being changed are uploaded.
4. Resource settings are updated, and a defer is registered to undo the change
   if subsequent steps fail.
5. A snapshot is taken.
6. The VCH is updated:
    a. If running, it is powered off.
    b. If volume stores are being added, those are created.
    c. The VCH itself is reconfigured.
    d. The VCH is powered on.
    e. We wait for the VCH to start and the service to be ready.
7. If the update is successful, we update the operations user.
8. If either the update or the operations user update is unsuccessful, we undo:
    a. We rollback to the snapshot we took.
    b. We delete uploaded images.
    c. We delete the snapshot we took.
    d. The defer undoes resource settings updates.

This process has a key characteristic: if a failure occurs, we cleanly return to
the initial state.

However, this process also has a key limitation: the process will fail at (6)(d)
or (6)(e) if the operations user requires additional privileges for the basic
operation of the VCH.

A simple solution to this might be to reverse the order of steps 6 and 7: update
the operations user before updating the VCH. This will address the limitation,
but will weaken the guarantee around cleanly returning to the initial state when
a failure occurs; it's not clear how to undo the changes to the operations user.

The effects of this are twofold:
1. Following the rollback of a change where additional privileges were granted
   to the operations user, those additional privileges are left as "cruft".
2. Following the rollback of a change where privileges were revoked from the
   operations user, the VCH will not start.

Options to improve this might include:
1. Looking into ways to undo changes to the operations user, perhaps following a
   similar pattern to the resource settings. This requires reading the old state
   from the system before making changes, including both the role-privilege map
   and resource-role associations.
2. Moving the operations logic in between (6)(c) and (6)(d) instead of between
   (5) and (6). This reduces the cases where we will leave cruft behind.

Reply:
I am not extremely familiar with this code. But I believe that we must determine what permissions the ops-user has before we begin to assemble the perms that are going to be needed in order to function. If we store that we could defere a configure back to those original permissions. @zjs why not make the ops-user changes before creating the volume-stores as well? They are created by vic-machine, but if we fail to assign the ops-user that will leave 1 less set of things to clean up. 

In the event of a failure what is the downside to returning the supplied ops-user to it's original permissions set?

Reply:
> In the event of a failure what is the downside to returning the supplied ops-user to it's original permissions set?

Returning the operations user to the original permissions set is the desired behavior.

> But I believe that we must determine what permissions the ops-user has before we begin to assemble the perms that are going to be needed in order to function. If we store that we could defere a configure back to those original permissions.

This is a possible approach, but actually extracting the "before" state is non-trivial. We need to track the changes being made to the privileges for each role as well as the roles being applied to each resource. Because of the way the vSphere APIs work, it's somewhat complicated to read all of that information.

> why not make the ops-user changes before creating the volume-stores as well? They are created by vic-machine, but if we fail to assign the ops-user that will leave 1 less set of things to clean up.

Currently, the operations user changes is the hardest thing to roll back (now tracked by #7814), so we do those as late in the process as possible.

Reply:
Added https://github.com/vmware/vic/issues/7814#issuecomment-383748290 to describe an alternate pattern, which is likely to be less work than tracking the old privileges/roles.